src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WorkerThread.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    internal static partial class ClrThreadPool
+    {
+        internal static class WorkerThread
+        {
+            private static LowLevelLifoSemaphore s_semaphore = new LowLevelLifoSemaphore(0, int.MaxValue);
+
+            private const int TimeoutMs = 20 * 1000;
+            
+            private static void WorkerThreadStart()
+            {
+                // TODO: Event: Worker Thread Start event
+
+                while (true)
+                {
+                    // TODO: Event:  Worker thread wait event
+                    while (s_semaphore.Wait(TimeoutMs))
+                    {
+                        if(!ThreadPoolWorkQueue.Dispatch())
```

Comment:
With that, when Dispatch returns true, it would be because it ran out of work and it would need to decrement numWorking before waiting. When Dispatch returns false, ShouldWorkerKeepRunning would have already decremented numWorking.